File: hier/project-euler/euler-002/euler2.cxx
// euler2.cxx - Project Euler exercise - Even Fibonacci numbers

#include <iostream>
#include <vector>

int main()
{
    constexpr long four_million {4L * 1000L * 1000L};
    
    // populate the sequence
    std::vector<long> a {1L, 2L};
    for (long back; (back = a.back()) <= four_million; )
        a.emplace_back(back + a.end()[-2]);

    // sum its even-valued elements
    long accum {0L};
    for (long k : a)
        if (k % 2L == 0L) accum += k;

    std::cout << accum << '\n';
}

File: hier/project-euler/euler-002/euler2.cs
// euler2.cs - Project Euler problem 2: Even Fibonacci numbers

using System;
using System.Collections.Generic;

[assembly:System.Reflection.AssemblyVersion("0.1.0.0")]

[Serializable] internal class UnreachableException : NotSupportedException { }

internal static class Program {
    private static IEnumerable<int> Fib(int a, int b)
    {
        for (; ; ) {
            yield return b;

            var tmp = a;
            a = b;
            b += tmp;
        }
    }

    private static int SumEvens(int max)
    {
        var n = 0;

        foreach (var k in Fib(1, 2)) {
            if (k > max) return n;

            if (k % 2 == 0) {
                n += k;
#if DEBUG
                Console.WriteLine("{0} ({1})", k, n);
#endif
            }
        }

        throw new UnreachableException();
    }

    private static void Main()
    {
        Console.WriteLine(SumEvens(4000000));
    }
}
